### ISSUE TITLE
[Request] (D-Bus) Interface to communicate with mdadm from other apps

### ISSUE THREAD

**Issue by Paiusco** [md-raid-utilities/mdadm#94]
I currently have a use-case where we can get the notification from the mdadm alert_cmd, so our application is notified that something happened with the disks (where we also parse the event)

**What we don't have is:** if after the faulty disk is replaced, we can tell mdadm that these partitions now should be added to the array without relying on the CLI.
Wouldn't be a nice addition to provide a D-Bus interface for this? So one app can mark disks as faulty/remove/add disks to the array without relying on CLI?

**Comment by mtkaczyk** [comment]
Hello @Paiusco 
I don't understand the problem here. Could you please present me scenario step by step. I need to understand about which device partitions we are talking here.

mdadm is not controlling partitions appearing and hiding, this all belongs to kernel and udev and it is a part of standard block layer functionality. MD device partitions should be presented during recovery (when faulty drive is being replaced), so I don't understand why you need some partitions to be added after finish of replacing the disk.

Probably I'm missing something and my comment could be wrong.

**Comment by Paiusco** [comment]
Thank you for your time @mtkaczyk 

Let's assume a RAID-1 with 2 disks, where one disk died. I configured an executable to notify my application whatever way I prefer. Great.

After that, I went to the DUT and swapped the failed disk to a new one, let's assume this new one already have the same partition table as the other disk already part of the RAID, although I still have to use the CLI to tell `mdadm` that these partitions need to be part of the RAID-1 now, right?

I'm assuming a workflow similar to [this](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/8/html/managing_storage_devices/managing-raid_managing-storage-devices#replacing-a-failed-disk-in-raid_managing-raid)

That's why my expectations were that one would be able to control these additions/removals through an API (e.g. D-Bus)

**Comment by mtkaczyk** [comment]
Ok, I think I got you. Thanks!

mdadm should clear partition table from the device before it is added to raid array via ioctl, (see remove_partitions() in mdadm). So, you can expect that partition table should disappear from the member drive before device is taken to raid array and rebuild is started.

First, please let me know what metadata you are using (native 0.9, native 1.2 or IMSM) and please provide me mdadm -E <device> from one of your member drives.

The problem with partitions visibility is not trivial and I don't want bother you with full context but in general whatever is visible under the member device if it is the part of raid array- it should not be used separately.
Races are always possible because it is extremally complex environment and partition detection still belongs to kernel, it may happen on any open but I don't remember details here.

There is also concept of hiding member devices but I'm not well familiar with this feature implementation.

**Comment by Paiusco** [comment]
Please take my terminology with a grain of salt in this case :)

When I said partitions, I meant the `/dev/sdX1`, as my RAID1 has 2 md devices (are actually 2), one for `/boot` and one for `/`.

So when I add a new disk into it in the DUT, assume I had already partitioned it, so I can just do `mdadm --manage /dev/md0 --add /dev/sdb1` and `mdadm --manage /dev/md1 --add /dev/sdb2`... and the mirroring just starts again on both block md devices.

> please let me know what metadata you are using

metadata=1.2

> provide me mdadm -E from one of your member drives

![image](https://github.com/user-attachments/assets/815eed1a-daf0-4304-815d-b1444e2b116c)

TL;DR: So my idea is that my own app can talk to mdadm to pick up the new plugged-in disks when I notify it in _any way_ besides command line, either through a lib, or D-Bus, as I stated. I there a way to do it?

**Comment by Paiusco** [comment]
UPDATE: I'm ~failing~ trying to play with the `POLICY` under config to try to make it do what I want without needing a controller app/script towards `mdadm`.

**Comment by mtkaczyk** [comment]
So, we have this feature but it is limited to real hardware devices, by-path link must exist.
You can probably hack it by creating by-path link for partition and configure mdadm but I cannot guarantee that it will work because mdadm may have some verifications. You have to try.

Raid autorebuilding is projected to work with disks/ pcie devices. If you want to enable it for partitions, we would need to find a solution.

**Comment by Paiusco** [comment]
Actually in my case it does still exist, the path is there:
![image](https://github.com/user-attachments/assets/7be127a5-1706-4a8c-aa4f-8038c045abd8)
the partitions are not virtual, at least not for this perspective:

![image](https://github.com/user-attachments/assets/ddbf6931-7c43-4c6c-b119-eb1cf2a3d13b)

Although I agree that this could have been done differently with only one RAID and the partitions inside. Do you think if this looked like:
sda
|
 -md0
  |
   -sda1
   -sda2
sdb
|
 -md0
  |
   -sdb1
   -sdb2

the autorebuilding would just work with a policy like:
`POLICY metadata=1.2 domain=test type=disk path=pci-0000:00:0d.0-ata* action=include` ?

_Note: the pci are from a VM, but I assume this has no issue for testing purposes_

**Comment by mtkaczyk** [comment]
> Actually in my case it does still exist, the path is there: ![image](https://private-user-images.githubusercontent.com/2374887/370712795-7be127a5-1706-4a8c-aa4f-8038c045abd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNzc0ODcsIm5iZiI6MTcyNzI3NzE4NywicGF0aCI6Ii8yMzc0ODg3LzM3MDcxMjc5NS03YmUxMjdhNS0xNzA2LTRhOGMtYWE0Zi04MDM4YzA0NWFiZDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTUxMzA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODdiOGU1NDdhZmYxYzg2NzkxNGJhNjY2M2U4YWZhMzM1NjU3MjRiZDFiOTMwZDc5NTAyMzZhOGI4ZDBhY2MwNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.p771a1212W31jbbJYRHNeUk4nwL7nZjS_ZvM07Ee2DU)

Indeed, I have no idea how I missed that, you are right. In that case, you should be able to configure autorebuild rules.
I see it now:

```
# ll /dev/disk/by-path/
total 0
drwxr-xr-x 2 root root 720 Sep 25 15:21 .
drwxr-xr-x 9 root root 180 Sep 25 15:21 ..
lrwxrwxrwx 1 root root   9 Sep 25 15:21 pci-0000:00:11.5-ata-5 -> ../../sda
lrwxrwxrwx 1 root root   9 Sep 25 15:21 pci-0000:00:11.5-ata-5.0 -> ../../sda
lrwxrwxrwx 1 root root  10 Sep 25 15:21 pci-0000:00:11.5-ata-5.0-part1 -> ../../sda1
lrwxrwxrwx 1 root root  11 Sep 25 15:21 pci-0000:00:11.5-ata-5.0-part10 -> ../../sda10
```

>the partitions are not virtual, at least not for this perspective:
> 
> ![image](https://private-user-images.githubusercontent.com/2374887/370713575-ddbf6931-7c43-4c6c-b119-eb1cf2a3d13b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNzc0ODcsIm5iZiI6MTcyNzI3NzE4NywicGF0aCI6Ii8yMzc0ODg3LzM3MDcxMzU3NS1kZGJmNjkzMS03YzQzLTRjNmMtYjExOS1lYjFjZjJhM2QxM2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTUxMzA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzU3OWExZmI5YTBjNWQ0MzQ0ZjliOWRhYjZmNzBjMjJmMDZhODlmMjZhNzJkNjBmZmFjMWZlNTM3ZjNkMzdjOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.WKCGSCUTWPIF38I-XEdMn4er1Uv9H3tlTORGoRJZelw)
> 

Right, partitions are not presented as virtual devices, MD devices are virtual:
```
 ll /sys/devices/virtual/block/
drwxr-xr-x  9 root root 0 Sep 25 15:21 md126
drwxr-xr-x  9 root root 0 Sep 25 15:21 md127
```
> Although I agree that this could have been done differently with only one RAID and the partitions inside. Do you think if this looked like: sda | |-md0 |||| ||||-sda1 ||||-sda2 sdb | |-md0 |||| ||||-sdb1 ||||-sdb2
> 
> the autorebuilding would just work with a policy like: `POLICY metadata=1.2 domain=test type=disk path=pci-0000:00:0d.0-ata* action=include` ?
> 
> _Note: the pci are from a VM, but I assume this has no issue for testing purposes_

We can explore both option as in both cases there are by-path links. I give it up to you. Naturally, option without raid on partitions would be simper but configuration should be similar.
In mdadm.conf you have define `POLICY` for every drive/partition so the simplest way is to do (this is what works for me with IMSM metadata):
`POLICY metadata=1.2 domain=test path=* action=spare-same-slot` 

Now, the drive you are attaching, if it doesn't contain metadata it must be `bare` drive (aka. brand-new, we expect zeros in first 4K and last 4K) with additional `bare drive rule` configured.
Mdadm is avoiding to take disks too aggressively. I did my best to describe it for IMSM here:
https://www.intel.com/content/dam/support/us/en/documents/memory-and-storage/linux-intel-vroc-userguide-333915.pdf
chapter 4.12.2, hope you will find it useful.

and here my script to simulate bare:
```
DEV=$1
mdadm --zero-superblock $DEV

# Zeroing first 4MiBi
dd if=/dev/zero of=$DEV bs=512 count=8 oflag=direct

# and last 4 MiBi
SIZE=$(blockdev --getsz $DEV)
dd if=/dev/zero of=$DEV bs=512 seek=$((SIZE - 4)) oflag=direct
```
It should work with partition at least theoretically :)
This feature is old and complicated, I will find some time tomorrow to simulate this on my setup and provide exact command.

**Comment by Paiusco** [comment]
Thanks for the detailed explanation, will try to update mine as well:

### Rebuild disks that have the same partition table
Let's say that I'll add a disk, and copy the partition table from the working disk into the new one.
After this has happened, the **only** way I was able to make an automatic sync occur was using `force-spare`, which trigger some other events that I don't really wanted, but if I try `spare-same-slot` or below, nothing happens. Am I crazy?

### For new disks
Nothing seems to happen, I tried wiping only the beginning and the end, but partly fails, so I just created a new empty disk on VM and still nothing. So I'm assuming for this case I need to recreate the partition table at least.

Extra question:
- When I create the POLICY, how does this end up as an udev rule? Do I have to do it manually? (e.g. running `mdadm --udev-rules > /etc/...` or is this done automatically (next boot) if I don't need a hotplug case? I couldn't find documentation mentioning something about it. On this case your document at Intel seems clearer to mention the reload of the rules, I'm considering updating the manpage

**Comment by Paiusco** [comment]
Currently, at work this was parked, and we'll work around the cases after a new bare disk is installed. So you may close this ticket, I'd still like that mdadm had a D-Bus interface to be able to talk to it. For now `[libblockdev](https://github.com/storaged-project/libblockdev/blob/master/src/plugins/mdraid.h)` will have to suffice.

I created https://github.com/md-raid-utilities/mdadm/pull/105 so `--udev-rules` is a bit more clear, and from there I could see that you can pass a file and it'll create it, if not, prints the rule on stdout.

> I just started my story with it (I'm maintainer starting from this year).
Congrats in being the new maintainer, hopefully it works fine for you!

**Comment by mtkaczyk** [comment]
Thank you!

Closing then.
